### PR TITLE
Import openssl-2.2.0

### PR DESCRIPTION
**Pull request by hsbt**
/cc @ioquatix

### CODE REVIEW

**Comment by ioquatix** [comment]
@hsbt thanks for your effort.

I just took a brief look at https://travis-ci.org/ruby/ruby/jobs/617114586 but I can’t see an OpenSSL related failure, do you mind checking that it worked/failed as expected?

**Comment by mame** [comment]
> I can’t see an OpenSSL related failure

See the log carefully, it is definitely related :sweat_smile: 

```
/home/travis/build/ruby/ruby/build/.ext/common/openssl.rb:13:in `require_relative': cannot load such file -- /home/travis/build/ruby/ruby/build/.ext/common/openssl.so (LoadError)
```

In Ruby master, openssl.so will be created in `.ext/<arch>/` (e.g., `.ext/x86_64-linux/`).  So we cannot use `require_relative` in this case

https://github.com/ruby/ruby/pull/2693/files#diff-8157890c77c799c399cd2ce08e2d63c0L13

File: ext/openssl/ossl_asn1.c [hunk 1]
```diff
@@ -1818,13 +1842,14 @@ do{\
...
     rb_define_alias(cASN1ObjectId, "long_name", "ln");
+    rb_define_method(cASN1ObjectId, "==", ossl_asn1obj_eq, 1);
     rb_attr(cASN1BitString, rb_intern("unused_bits"), 1, 1, 0);
 
     rb_define_method(cASN1EndOfContent, "initialize", ossl_asn1eoc_initialize, 0);
     rb_define_method(cASN1EndOfContent, "to_der", ossl_asn1eoc_to_der, 0);
 
     class_tag_map = rb_hash_new();
-    rb_gc_register_mark_object(class_tag_map);
+    rb_global_variable(&class_tag_map);
```
Comment (nobu) [inline ext/openssl/ossl_asn1.c]: This reverts the fix for https://bugs.ruby-lang.org/issues/16196, and will cause a segfault.

Comment (ioquatix) [reply ext/openssl/ossl_asn1.c]: Okay so it was never upstreamed? I can pull this into OpenSSL source code. Should we check for other issues like this?

Comment (nobu) [reply ext/openssl/ossl_asn1.c]: There may be other fixes to be ported. 😭

File: ext/openssl/ossl_ssl.c [hunk 2]
```diff
@@ -592,7 +592,7 @@ ssl_renegotiation_cb(const SSL *ssl)
 #if !defined(OPENSSL_NO_NEXTPROTONEG) || \
     defined(HAVE_SSL_CTX_SET_ALPN_SELECT_CB)
 static VALUE
-ssl_npn_encode_protocol_i(RB_BLOCK_CALL_FUNC_ARGLIST(cur, encoded))
+ssl_npn_encode_protocol_i(VALUE cur, VALUE encoded)
```
Comment (nobu) [inline ext/openssl/ossl_ssl.c]: This reverts 3cae73133cf, and causes compile error on mswin.

**Comment by ioquatix** [comment]
Thanks @mame I will pull all changes upstream.

File: ext/openssl/ossl_asn1.c (hunk 1 above)
Comment (ioquatix) [reply ext/openssl/ossl_asn1.c]: Okay, I will upstream all these changes.

**Comment by ioquatix** [comment]
Okay I upstreamed all changes to `openssl` repository:

https://github.com/ruby/openssl/compare/f48885c9400c272fa2647c47a891a515b97cda59...master

**Comment by ioquatix** [comment]
I merged all changes here downstream. What is the process to bring it upstream, so I can check if I missed anything?

**Comment by hsbt** [comment]
@ioquatix At least, you should resolve the test failures on, Windows on GitHub Actions, Travis and AppVeyor. After that, I will merge this into ruby/ruby.

**Comment by hsbt** [comment]
@ioquatix 

I did update with the current master branch of ruby/openssl. But it was failed with compile phase caused by `ossl_version.h`.

**Comment by ioquatix** [comment]
Thanks @hsbt for all your effort.

Can you explain to me how to do this? I assume there is some command to pull the source from the openssl repo into ruby master. If I can do this, I can avoid using your time and try to fix most of the issues myself.

**Comment by ioquatix** [comment]
Thanks for this. I didn’t expect changing `VERSION` constant would cause so many issues.

**Comment by hsbt** [comment]
@ioquatix mswin build still failed at https://github.com/ruby/ruby/pull/2693/checks?check_run_id=426744656 and mingw tests also failed https://github.com/ruby/ruby/pull/2693/checks?check_run_id=426744588.

We need to fix if for releasing openssl-2.2.0.

**Comment by jdhollis** [comment]
I don't currently have a Windows box to work with, so I can't `git bisect` my way to victory.

However, while trying to figure out what changes might be involved, I came across this commit mentioning `no OPENSSL_Applink`: https://github.com/ruby/ruby/commit/df94c66f71448cf30b34375349fd201d1d035423

I don't know much about compiling software on Windows these days, but here's an OpenSSL FAQ that seems to address the issue: https://www.openssl.org/docs/faq.html#PROG3

So it looks like we're not linking in the appropriate shim for some reason.

I'll keep digging.

**Comment by jdhollis** [comment]
For the [MinGW failure](https://github.com/ruby/ruby/pull/2693/checks?check_run_id=426744588), I think we need to make certain that [ruby/.github/workflows/mingw.yml](https://github.com/ruby/ruby/blob/master/.github/workflows/mingw.yml) aligns with [openssl/.github/workflows/workflow.yml](https://github.com/ruby/openssl/blob/master/.github/workflows/workflow.yml).

It looks like @MSP-Greg was responsible for both, so perhaps he'd be the best person to weigh in.

**Comment by jdhollis** [comment]
The other three failures seem to be related to this commit: https://github.com/ruby/openssl/commit/debaca25604c7eee40b4f930ab9bcb2dcf4e18eb

From a cursory search, it's not clear to me that you can use zero size arrays with Visual Studio, but it does appear that we can fix the [Travis failure](https://travis-ci.org/ruby/ruby/builds/646235843?utm_source=github_status&utm_medium=notification) by adding the appropriate flags: https://travis-ci.org/ruby/ruby/jobs/646235849#L2037

**Comment by MSP-Greg** [comment]
I'll have look soon.

**Comment by ioquatix** [comment]
All tests are passing in main repo: https://github.com/ruby/openssl/actions/runs/33674529

**Comment by movitto** [comment]
@jdhollis @MSP-Greg @ioquatix Unfortunately I don't have a visual studio / mingw environment to test locally on but I can send a followup PR changing the 'sizer' buffer size to 1 to mitigate this issue. Thoughts?

**Comment by ioquatix** [comment]
> @jdhollis @MSP-Greg @ioquatix Unfortunately I don't have a visual studio / mingw environment to test locally on but I can send a followup PR changing the 'sizer' buffer size to 1 to mitigate this issue. Thoughts?

Can't we just use a null pointer?

**Comment by MSP-Greg** [comment]
I'm not a strong C type, no idea about the mswin/msvc issue.

I'm going to try building locally, MinGW first...

**Comment by jdhollis** [comment]
From [openssl/ssl_lib.c](https://github.com/openssl/openssl/blob/master/ssl/ssl_lib.c#L1431):

```c
/* return length of latest Finished message we sent, copy to 'buf' */
size_t SSL_get_finished(const SSL *s, void *buf, size_t count)
{
    size_t ret = 0;

    ret = s->s3.tmp.finish_md_len;
    if (count > ret)
        count = ret;
    memcpy(buf, s->s3.tmp.finish_md, count);
    return ret;
}

/* return length of latest Finished message we expected, copy to 'buf' */
size_t SSL_get_peer_finished(const SSL *s, void *buf, size_t count)
{
    size_t ret = 0;

    ret = s->s3.tmp.peer_finish_md_len;
    if (count > ret)
        count = ret;
    memcpy(buf, s->s3.tmp.peer_finish_md, count);
    return ret;
}
```

My C is rusty enough that it's not clear to me what could end up being copied to `buf`.

**Comment by movitto** [comment]
> > @jdhollis @MSP-Greg @ioquatix Unfortunately I don't have a visual studio / mingw environment to test locally on but I can send a followup PR changing the 'sizer' buffer size to 1 to mitigate this issue. Thoughts?
> 
> Can't we just use a null pointer?

:+1: just sent a PR (see above). Tests pass locally, let me know if this suffices!

**Comment by jdhollis** [comment]
That feels wrong to me. Isn't the behavior of `memcpy` to `NULL` undefined?

**Comment by movitto** [comment]
> That feels wrong to me. Isn't the behavior of `memcpy` to `NULL` undefined?

Ya reading the documenation a bit more, I'm thinking that I'll have to roll back this latest PR. Passing NULL to memcpy is undefined behaviour. Array of length 1 is still an option that's on the table though.

**Comment by jdhollis** [comment]
Ah. That said, what happens when you pass 0 into `memcpy` for byte count? Maybe it's just a noop?

**Comment by jdhollis** [comment]
According to this, it might be fine: https://stackoverflow.com/questions/3751797/can-i-call-memcpy-and-memmove-with-number-of-bytes-set-to-zero

Damn, my C is rusty 😛

**Comment by jdhollis** [comment]
Now that I think about it, I believe the behavior is still undefined even if the count is 0. So maybe a 1-byte buffer is best.

**Comment by MSP-Greg** [comment]
I just built MinGW locally from this PR, and it passed.

Also, @movitto's PR https://github.com/ruby/openssl/pull/315 passed on the MinGW build using a Ruby master build that is the current commit.  See: https://github.com/movitto/openssl/runs/428569780#step:6:10

Confused.

Also `ruby -ropenssl -e "puts OpenSSL::VERSION"` returns
`uninitialized constant OpenSSL::VERSION (NameError)`

**Comment by ioquatix** [comment]
https://github.com/ruby/openssl/pull/315

Please discuss PR specific things on the relevant PR :)

I keep on jumping between issues haha.

**Comment by MSP-Greg** [comment]
1. Please see PR https://github.com/ruby/ruby/pull/2886, as I think it needs to be accepted and this rebased for MinGW to pass.

2. I added @movitto's PR https://github.com/ruby/openssl/pull/315 to this and built in my fork.  mswin compiled (it's currently failing), but then it has the same test error as MinGW.  Haven't tracked it down yet...

**Comment by movitto** [comment]
I opened a seperate PR w/ the 1-byte array fix, see https://github.com/ruby/openssl/pull/316

My gut is to go with this later one as the solution as it is safest and least ambiguous, eg least likely to cause unforseen issues in some random environment / platform where NULL pointers passed to memcpy cause issues.

Test pass locally, again don't have access to a visual studio, mingw env to test on.

**Comment by MSP-Greg** [comment]
More info about the mswin failure at https://github.com/ruby/openssl/issues/318

**Comment by hsbt** [comment]
The test was passed. We can merge this and release openssl-2.2.0 with the following pull-requests.

 * https://github.com/ruby/openssl/pull/320
* https://github.com/ruby/openssl/pull/321

**Comment by hsbt** [comment]
@rhenium I did merge this branch for testing with the ruby repo. Can you add the commits before releasing openssl-2.2.0? Thanks.

**Comment by jdhollis** [comment]
Following along with this release, it does seem to me like it might be useful to establish a firmer process around how commits flow from `ruby/openssl` to `ruby/ruby` (or perhaps from `ruby/ruby` to `ruby/openssl` given the impact on the larger build).

**Pull request merged**